rethinkdb: Initial Commit

meenie opened the pull request.

Commit:
rethinkdb-cluster: Initial Commit

Commit:
Fix chart name.

Commit:
Make the NOTES.txt file more accurate.

Commit:
Use k8s 1.6 storageClassName notation for PVC.

Commit:
RethinkDB directory needs to be initially empty.

When using a provisioned volume from something like GCE, the mounted directory has a  folder inside of it. This breaks RethinkDB when initializing.

Commit:
Add an extra k8s script to create a StorageClass on GCE.

unguiculus commented on incubator/rethinkdb-cluster/README.md:
```diff
@@ -0,0 +1,131 @@
+| `cluster.podAnnotations`                | Annotations to be added to RethinkDB Cluster Pods                                  | `{}`                                                |
+| `cluster.serviceAnnotations`            | Annotations to be added to RethinkDB Cluster Service                               | `{}`                                                |
+| `cluster.persistentVolume.enabled`      | If `true`, persistent volume claims are created                                    | `true`                                              |
+| `cluster.persistentVolume.storageClass` | Persistent volume storage class                                                    | `volume.alpha.kubernetes.io/storage-class: default` |
```
I know this is done in many charts, but does it still make sense to make the alpha annotation the default, given the fact that Kubernetes 1.5+ is a prerequisite?

meenie replied:
I actually did start using `storageClassName` in the PVC.  I also believe that's k8s 1.6+ as well. So I think I should change the pre-reqs to 1.6+. Thoughts?

unguiculus replied:
I think the test system for the charts is still on 1.5.

unguiculus commented on incubator/rethinkdb-cluster/README.md:
```diff
@@ -0,0 +1,131 @@
+
+## Acknowledgment of Previous Works
+
+I have heavely barrowed and extended code (peer discovery and probe) from the following project to build this Helm Chart and Docker image: https://github.com/rosskukulinski/kubernetes-rethinkdb-cluster.
```
`heavely barrowed` -> `heavily borrowed`

unguiculus commented on incubator/rethinkdb-cluster/extra/create-gce-storageclass.yaml:
```diff
@@ -0,0 +1,7 @@
+kind: StorageClass
```
Not sure if it is necessary to supply this manifest. It is fairly standard stuff.  At least, I would mention its purpose in the readme file.

meenie replied:
👍

unguiculus commented on incubator/rethinkdb-cluster/templates/rethinkdb-service.yaml:
```diff
@@ -0,0 +1,44 @@
+  selector:
+    app: "{{ template "name" . }}-cluster"
+    release: {{ .Release.Name | quote }}
+---
```
I'd prefer to have this in two separate files.

meenie replied:
👍

unguiculus commented on incubator/rethinkdb-cluster/templates/rethinkdb-service.yaml:
```diff
@@ -0,0 +1,44 @@
+{{ toYaml .Values.proxy.serviceAnnotations | indent 4 }}
+  {{- end }}
+spec:
+  type: LoadBalancer
```
Please make service type etc. configurable. Here's a good example of how this can be done.
https://github.com/kubernetes/charts/blob/master/stable/nginx-ingress/templates/controller-service.yaml#L16

meenie replied:
Do you not need to explicitly set a `type`?

unguiculus replied:
You have to set a type, but not everyone may want to expose it with a load balancer. You could e. g. use an ingress controller and make the service type `ClusterIP`. Or you may not want to expose it at all. Depends on your use case.

meenie replied:
👍

unguiculus reviewed (changes requested):
Looks great. I made a few comments.

meenie commented:
@unguiculus Do you think this should be called just `rethinkdb`?  I assume no one would want to use RethinkDB without using it as a cluster...

unguiculus commented:
I think `rethinkdb` would be better. Also, I think a chart qualifies to dwell under `stable` if it does not use any alpha APIs. So, you may want to move it there.

Commit:
Addressing PR Comments.

Commit:
Rename to Recursively removing directory /Users/codyl/dev-files/charts/rethinkdb_data/tmp
Initializing directory /Users/codyl/dev-files/charts/rethinkdb_data
Running rethinkdb 2.3.4 (CLANG 7.3.0 (clang-703.0.31))...
Running on Darwin 16.6.0 x86_64
Loading data from directory /Users/codyl/dev-files/charts/rethinkdb_data and move to stable.

Commit:
Shorten down .gitignore file.

meenie commented:
@unguiculus I've made the requested changes, renamed to just `rethinkdb` and moved into `stable`.

Commit:
Make using the RethinkDB Admin Console easier.

Commit:
Fix some wrong names/paths.

Commit:
Due to a bug with secrets, don't auto-gen a password.

https://github.com/kubernetes/charts/issues/980

Commit:
Remove extra file.

prydonius commented on stable/rethinkdb/Chart.yaml:
```diff
@@ -0,0 +1,13 @@
+- rethinkdb
+- database
+- nosql
+home: https://www.rethinkdb.com
```
Can we get a rethinkdb logo so it shows up pretty on http://kubeapps.com/?

meenie replied:
[Which one do you suggest?](https://www.google.com/search?newwindow=1&site=&tbm=isch&source=hp&biw=2560&bih=1316&q=rethinkdb+logo&oq=rethinkdb+logo&gs_l=img.3..35i39k1.1077.2762.0.2925.15.15.0.0.0.0.136.1063.12j2.14.0....0...1.1.64.img..1.14.1056.0..0j0i24k1.IpcYB2a9CwM). I suppose the one with the "R" in a square?

thomasmodeneis replied:
I think the best place to pick a icon from is inside the repo -> https://github.com/rethinkdb/docs/tree/master/_jekyll/_images/api_illustrations it all released under https://creativecommons.org/licenses/by-sa/4.0/

meenie replied:
Using this icon: http://i.imgur.com/bZ75N6r.jpg

prydonius commented on stable/rethinkdb/templates/NOTES.txt:
```diff
@@ -0,0 +1,20 @@
+NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+      Watch the status with: 'kubectl get svc --namespace {{ .Release.Namespace }} -w {{ template "fullname" . }}-proxy'
+
+  echo Host: $(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "fullname" . }}-proxy -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
```
It would be good to have different instructions based on the `service.type` value, like we have in other charts

meenie replied:
👍

prydonius commented on stable/rethinkdb/templates/rethinkdb-cluster-service.yaml:
```diff
@@ -0,0 +1,22 @@
+    heritage: {{ .Release.Service | quote }}
+    release: {{ .Release.Name | quote }}
+  annotations:
+    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
```
I can't find any documentation about this, if it requires alpha support then this will need to be in incubator.

unguiculus replied:
See https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L62

This is actually used on several stable charts.

prydonius replied:
Ah okay, thanks @unguiculus - I assume the "alpha" is just left behind and so this isn't actually an alpha feature.

prydonius commented on stable/rethinkdb/templates/rethinkdb-secrets.yaml:
```diff
@@ -0,0 +1,12 @@
+    heritage: "{{ .Release.Service }}"
+type: Opaque
+data:
+  rethinkdb-password: {{ .Values.rethinkdbPassword | b64enc | quote }}
```
The password should be randomly generated by default

meenie replied:
I actually had it like that before, but there is an issue when you do that: https://github.com/kubernetes/charts/issues/980. I'm not really sure what would be the best way to combat that issue.

unguiculus replied:
ping @prydonius

thomasmodeneis replied:
that looks like a bit of a blocker, and so far with no solution. 
@prydonius is this a blocker this chart to be released ? how are other databases like cassandra/mongodb dealing with the issue ?

prydonius replied:
sorry for the slow turnaround on this, IMO we should be consistent with the other databases for now until we come up with a better solution. You can still use the value if it's specified so I don't think this will change your usecase?

prydonius replied:
Looks like some other database don't use random passwords by default (MariaDB, MongoDB) as @thomasmodeneis found, but those should really be updated to use them. A secure default is really important.

meenie replied:
@prydonius I agree there should be a secure default. But the issue people can run into is if they do a `helm upgrade` in the future, it will generate a new password unless they put the first password that was auto-generated in their local `values.yaml` file they are using to configure the k8s deployment.  But if that's an acceptable risk, then I'm all for auto-generating :).

thomasmodeneis replied:
I don't really think a secure default is of any importance here.
We assume that the person taking care of the databases, will as well take care to generate the secure password and protect the cluster behind a firewall.

As we see on all other databases(MariaDB, MongoDB, Redis, CochroachDB, etc), the default is "do the simple" and generate a default password due to the problem on `helm upgrade`.

prydonius replied:
@thomasmodeneis Whilst a serious admin will take care to generate a secure password and protect the cluster, it's always safer to set secure defaults for an installation. [MongoDB](https://www.theregister.co.uk/2017/01/09/mongodb/) is a great example of where not using secure defaults can go horribly wrong.

The `helm upgrade` problem is an issue, but something we need to push for fixing upstream in Helm, until then we should document the upgrade procedure and explain that the password needs to be set when upgrading.

If we feel strongly about not randomly generating a password because of this issue, we definitely need to set the default service type to ClusterIP (however I think doing both is best).

meenie replied:
@prydonius @unguiculus I like the compromise of setting the default service type for the proxy to `ClusterIP`.  If I do that and leave the password as it is, would that satisfy this issue?

prydonius replied:
@meenie sounds good to me - thanks

Commit:
Addressing PR comments and some other changes

 - Updated NOTES.txt to include more tailored instructions based on proxy service type.
 - Added an optional StorageClass.
 - Simplified rethink config to not use a ConfigMap.
 - Added optional RethinkDB Driver TLS configuration.
 - Updated README.md to reflect new config changes.

Commit:
Tiny fix in README.md.

meenie commented:
@unguiculus and @prydonius: I've made the requested changes.  With regards to auto-generating a password, I still see no way around the issue with generating a new password each time you upgrade.

thomasmodeneis commented:
Hi,
I was also checking other databases charts and found that MongoDB is doing actually the very similar thing:
```
  mongodb-root-password: {{ default "" .Values.mongodbRootPassword | b64enc | quote }}
  mongodb-password: {{ default "" .Values.mongodbPassword | b64enc | quote }}
```

https://github.com/kubernetes/charts/blob/master/stable/mongodb/templates/secrets.yaml

prydonius commented on stable/rethinkdb/Chart.yaml:
```diff
@@ -0,0 +1,14 @@
+- https://hub.docker.com/r/codylundquist/helm-rethinkdb-cluster/
+icon: http://i.imgur.com/bZ75N6r.jpg
+maintainers:
+- name: Cody Lundquist
```
can you update the name to `meenie`, we're now switching to github ids

meenie replied:
👍

prydonius reviewed (changes requested):
sorry for taking some time to get back to this

Commit:
Use github username for Chart.yaml.

unguiculus commented:
+1 for setting the default service type to ClusterIP

Commit:
Default to ClusterIP for proxy service.

meenie commented:
@unguiculus @prydonius: I've changed the default to use `ClusterIP` instead of `LoadBalancer`.

prydonius commented:
@k8s-bot ok to test

prydonius reviewed (approved):
Thanks for the great contribution @meenie :), excited to see this in.

@thomasmodeneis thank you for helping with the review!
